### PR TITLE
feat(JOML): migrate BaseFacet2D for BlockArea

### PR DESCRIPTION
**Pull request by pollend**
Here is the start of a migration for BaseFace2D to take advantage of the BlockArea change. 

some 2d facets needs: https://github.com/MovingBlocks/Terasology/pull/4339

- [ ] https://github.com/Terasology/AnotherWorld/pull/18
- [x] https://github.com/Terasology/Cities/pull/43
- [ ] https://github.com/Terasology/CoreWorlds/pull/21
- [ ] https://github.com/Terasology/DynamicCities/pull/82
- [ ] https://github.com/Terasology/Exoplanet/pull/16
- [ ] https://github.com/Terasology/GooeyDefence/pull/52
- [ ] https://github.com/Terasology/Inferno/pull/18
- [ ] https://github.com/Terasology/Lakes/pull/17
- [ ] https://github.com/Terasology/LightAndShadow/pull/182
- [ ] https://github.com/Terasology/MetalRenegades/pull/128
- [ ] https://github.com/Terasology/ModuleTestingEnvironment/pull/33
- [ ] https://github.com/Terasology/ParadIce/pull/8
- [ ] https://github.com/Terasology/PolyWorld/pull/32
- [ ] https://github.com/Terasology/ShatteredPlanes/pull/16
- [ ] https://github.com/Terasology/StaticCities/pull/20
- [ ] https://github.com/Terasology/SurfaceFacets/pull/8
- [ ] https://github.com/Terasology/WeatherManager/pull/19

<details>
<summary>Checkout all related PRs with <code>gh</code></summary>

```bash
gh pr checkout 4364
pushd modules
pushd AnotherWorld; gh pr checkout 18; popd;
pushd CoreWorlds; gh pr checkout 21; popd;
pushd DynamicCities; gh pr checkout 82; popd;
pushd Exoplanet; gh pr checkout 16; popd;
pushd GooeyDefence; gh pr checkout 52; popd;
pushd Inferno; gh pr checkout 18; popd;
pushd Lakes; gh pr checkout 17; popd;
pushd LightAndShadow; gh pr checkout 182; popd;
pushd MetalRenegades; gh pr checkout 128; popd;
pushd ModuleTestingEnvironment; gh pr checkout 33; popd;
pushd ParadIce; gh pr checkout 8; popd;
pushd PolyWorld; gh pr checkout 32; popd;
pushd ShatteredPlanes; gh pr checkout 16; popd;
pushd StaticCities; gh pr checkout 20; popd;
pushd SurfaceFacets; gh pr checkout 8; popd;
pushd WeatherManager; gh pr checkout 19; popd;
popd
```

</details>

### CODE REVIEW

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet3D.java [hunk 1]
```diff
@@ -29,8 +29,8 @@
 
     public BaseBooleanFieldFacet3D(BlockRegion targetRegion, Border3D border) {
         super(targetRegion, border);
-        Vector3i size = getRelativeRegion().getSize(new Vector3i());
-        data = new boolean[size.x * size.y * size.z];
+        data =
+            new boolean[getRelativeRegion().getSizeX() * getRelativeRegion().getSizeY() * getRelativeRegion().getSizeZ()];
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet3D.java:33]:
```suggestion
            new boolean[getRelativeRegion().volume()];
```

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseFieldFacet2D.java [hunk 2]
```diff
@@ -29,10 +26,9 @@
 
     private float[] data;
 
-    public BaseFieldFacet2D(BlockRegion targetRegion, Border3D border) {
+    public BaseFieldFacet2D(BlockRegionc targetRegion, Border3D border) {
         super(targetRegion, border);
-        Vector2i size = getRelativeRegion().size();
-        this.data = new float[size.x * size.y];
+        this.data = new float[getRelativeRegion().getSizeX() * getRelativeRegion().getSizeX()];
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/facets/base/BaseFieldFacet2D.java:31]:
```suggestion
        this.data = new float[getRelativeRegion().getSizeX() * getRelativeRegion().getSizeY()];
```

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java [hunk 3]
```diff
@@ -31,8 +31,7 @@
 
     public BaseBooleanFieldFacet2D(BlockRegion targetRegion, Border3D border) {
         super(targetRegion, border);
-        Vector2i size = getRelativeRegion().size();
-        data = new boolean[size.x * size.y];
+        data = new boolean[getRelativeRegion().getSizeX() * getRelativeRegion().getSizeY()];
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java:34]: I'm somewhat confused - don't we typically treat a 2D variant of some region as projection onto the XZ plane? Should this have been `size.x * size.z` before? 🤔

**Review by skaldarnar — CHANGES_REQUESTED** [review]

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java (hunk 3 above)
Comment (pollend) [reply engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java:34]: you might of found a bug?

Comment (pollend) [reply engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java:34]: nope getRelativeRegion() is a BlockArea so its x and y.

File: engine/src/main/java/org/terasology/world/generation/WorldFacet2D.java [hunk 4]
```diff
@@ -24,10 +24,10 @@
     /**
      * @return The region of the world covered by this facet
      */
-    Rect2i getWorldRegion();
+    BlockAreac getWorldRegion();
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/WorldFacet2D.java:27]: I think we should rename this to `getWorldArea`. I would like to reserve `region` for 3D block regions, and `area` for 2d block areas.
```suggestion
    BlockAreac getWorldArea();
```

File: engine/src/main/java/org/terasology/world/generation/WorldFacet2D.java [hunk 5]
```diff
@@ -24,10 +24,10 @@
...
 
     /**
      * @return The region covered by this facet, relative to the target region
      */
-    Rect2i getRelativeRegion();
+    BlockAreac getRelativeRegion();
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/WorldFacet2D.java:32]: Same here:
```suggestion
    BlockAreac getRelativeArea();
```

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java [hunk 6]
```diff
@@ -31,8 +31,7 @@
 
     public BaseBooleanFieldFacet2D(BlockRegion targetRegion, Border3D border) {
         super(targetRegion, border);
-        Vector2i size = getRelativeRegion().size();
-        data = new boolean[size.x * size.y];
+        data = new boolean[getRelativeRegion().area()];
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java:34]: That's way better, but as commented on the interface we should rename the method:
```suggestion
        data = new boolean[getRelativeArea().area()];
```

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseFacet2D.java [hunk 7]
```diff
@@ -18,43 +18,46 @@
...
  */
 public class BaseFacet2D implements WorldFacet2D {
 
-    private Rect2i worldRegion;
-    private Rect2i relativeRegion;
+    private BlockArea worldRegion;
+    private BlockArea relativeRegion;
 
-    public BaseFacet2D(BlockRegion targetRegion, Border3D border) {
+    public BaseFacet2D(BlockRegionc targetRegion, Border3D border) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/facets/base/BaseFacet2D.java:35]: We should definitely add some JavaDoc here stating how the 3D block region `targetRegion` is translated into a 2D area. In particular, how the projection down to 2D works...

**Review by skaldarnar — CHANGES_REQUESTED** [review]
Sorry that we need to do another round here, but I think we should try to get the names right while we are at it!

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java (hunk 3 above)
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/world/generation/facets/base/BaseBooleanFieldFacet2D.java]: alright, then this also boils down to renaming that method `getRelativeArea()` 😉

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseFacet2D.java (hunk 7 above)
Comment (pollend) [reply engine/src/main/java/org/terasology/world/generation/facets/base/BaseFacet2D.java:35]: we could always do with more javadocs

File: engine/src/main/java/org/terasology/world/generation/facets/base/BaseSparseFacet2D.java [hunk 8]
```diff
@@ -27,33 +28,33 @@
  * A base class for sparse (map-based) 2D facets.
  */
 public abstract class BaseSparseFacet2D implements WorldFacet2D {
-    private Rect2i worldRegion;
-    private Rect2i relativeRegion;
+    private BlockAreac worldRegion;
+    private BlockAreac relativeRegion;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/facets/base/BaseSparseFacet2D.java:32]:
```suggestion
    private BlockAreac worldArea;
    private BlockAreac relativeArea;
```

**Review by skaldarnar — CHANGES_REQUESTED** [review]
let's also rename the internal field names, please.

**Review by skaldarnar — DISMISSED** [review]

**Review by skaldarnar — APPROVED** [review]

**Pull request merged**